File: IdeaProjects/Meeting System/src/models/MeetingSchedulingSystem.java
package models;

import java.util.List;

public class MeetingSchedulingSystem {
    private MeetingSchedular schedular;
    private List<User> users;
    private List<MeetingRoom> room;

    private static MeetingSchedulingSystem meetingSchedulingSystemInstance = null;
    public MeetingSchedulingSystem() {}

    public static MeetingSchedulingSystem getInstance() {
        if(meetingSchedulingSystemInstance == null){
            meetingSchedulingSystemInstance = new MeetingSchedulingSystem();
        }
        return meetingSchedulingSystemInstance;
    }

    public void addRoom(){
        System.out.println("Added In Room!");
    }

    public void removeRoom(){
        System.out.println("Remove From Room!");
    }


    public void scheduleMeeting(List<User> users, Interval interval){
        schedular.scheduleMeeting();
    }

    public void cancelingMeeting(){
        schedular.cancelMeeting();
    }
}

File: 09_Design Problems/04_Meeting Schedular System/Mian Class/MeetingSchedulingSystem.cpp
#include <iostream>
#include <list>
#include "../Models/User.cpp"
#include "../Models/MeetingRoom.cpp"
#include "../Models/Invterval.cpp"
#include "../Models/MeetingSchedular.cpp"

using namespace std;

class MeetingSchedulingSystem {
    MeetingSchedular schedular;
    list<User> users;
    list<MeetingRoom> meetingRooms;

    // signle instance of meeting schedular
    static MeetingSchedulingSystem* instance;

    MeetingSchedulingSystem();

    public:

        // add threadsafety for singleton instance work properly in multi-threaded environment
        static MeetingSchedulingSystem* getInstance() {
            if(instance == nullptr) {
                instance = new MeetingSchedulingSystem();
            }
            return instance;
        }

        void addRoom() {
            cout << "Adding meeting room to system" << endl;
        }

        void removeRoom(MeetingRoom room) {
            cout << "Removing meeting room to system" << endl;
        }

        void scheduleMeeting(list<User> attendees, MeetingRoom room, Interval interval) {
            schedular.scheduleMeeting(attendees, room, interval);
            cout << "Scheduling meeting" << endl;
        }

        void cancelMeeting() {
            cout << "Cancelling meeting" << endl;
        }
};